### ISSUE TITLE
Contrastive train loss experiment results

### ISSUE THREAD

**Issue by hominot** [hominot/research#48]
cub200
contrastive
alpha=4
group_size=4
batch_size=64
learning_rate=3e-05

Importance Sampling

epoch 0

0.9110286607328021
recall@1: 0.3262997974341661
recall@2: 0.4464888588791357
recall@4: 0.5678595543551654
recall@8: 0.6968264686022958
auc: 0.67897
nmi: 0.44501778801166497

epoch 1

0.2528889923969991
recall@1: 0.27819041188386223
recall@2: 0.38723835246455096
recall@4: 0.5158676569885212
recall@8: 0.6551316677920324
auc: 0.71883
nmi: 0.41074657689818894

epoch 2

0.1602272068851267
recall@1: 0.3779540850776502
recall@2: 0.5047265361242403
recall@4: 0.6438217420661715
recall@8: 0.761985145172181
auc: 0.83002
nmi: 0.5167311850166868

epoch 3

0.14522254866930662
recall@1: 0.424713031735314
recall@2: 0.5496286293045239
recall@4: 0.6775827143821742
recall@8: 0.7905131667792032
auc: 0.86043
nmi: 0.5536802680617331

epoch 4

0.1362596434130376
recall@1: 0.4529034436191762
recall@2: 0.5823767724510466
recall@4: 0.7034098582039163
recall@8: 0.8146522619851452
auc: 0.87855
nmi: 0.5849511140849519

epoch 5

0.12944526673625284
recall@1: 0.4665766374071573
recall@2: 0.5945307224848075
recall@4: 0.7191087103308575
recall@8: 0.824274139095206
auc: 0.88789
nmi: 0.5891073879179346

epoch 6

0.12436806991056022
recall@1: 0.47805536799459825
recall@2: 0.6095543551654288
recall@4: 0.7307562457798784
recall@8: 0.8340648210668468
auc: 0.89279
nmi: 0.6030809332057642

epoch 7

0.12170680653092435
recall@1: 0.4962862930452397
recall@2: 0.6299797434166104
recall@4: 0.7454422687373397
recall@8: 0.8416610398379474
auc: 0.8973
nmi: 0.6096622580595261

epoch 8

0.11803852240952614
recall@1: 0.5057393652937204
recall@2: 0.6350438892640108
recall@4: 0.7422349763673194
recall@8: 0.8364280891289669
auc: 0.89511
nmi: 0.6062770727874947

epoch 9

0.11453363682536342
recall@1: 0.512322754895341
recall@2: 0.6406144496961512
recall@4: 0.7538825118163404
recall@8: 0.8448683322079676
auc: 0.89582
nmi: 0.6094918150282906

epoch 10

0.11177368136150546
recall@1: 0.513166779203241
recall@2: 0.6396016205266711
recall@4: 0.7525320729237002
recall@8: 0.8436866981769074
auc: 0.90154
nmi: 0.6115236003883928

epoch 11

0.10950312901949218
recall@1: 0.5015192437542201
recall@2: 0.6291357191087104
recall@4: 0.7481431465226198
recall@8: 0.8414922349763673
auc: 0.90103
nmi: 0.6133199282146362

epoch 12

0.10788239969842635
recall@1: 0.5185685347738015
recall@2: 0.6455097906819717
recall@4: 0.7565833896016205
recall@8: 0.8463875759621877
auc: 0.9022
nmi: 0.6098058845732562

epoch 13

0.10564851392774151
recall@1: 0.5180621201890614
recall@2: 0.6439905469277515
recall@4: 0.7537137069547603
recall@8: 0.8436866981769074
auc: 0.9009
nmi: 0.611726922784107

epoch 14

0.10378515831530596
recall@1: 0.5199189736664416
recall@2: 0.6485482781904118
recall@4: 0.7554017555705604
recall@8: 0.8522957461174882
auc: 0.90067
nmi: 0.612552894424817

epoch 15

0.10281045647559497
recall@1: 0.5275151924375422
recall@2: 0.6534436191762323
recall@4: 0.7665428764348413
recall@8: 0.8517893315327482
auc: 0.90475
nmi: 0.6168086238013726

epoch 16

0.10095393380333818
recall@1: 0.5241390952059419
recall@2: 0.6525995948683322
recall@4: 0.7662052667116813
recall@8: 0.8460499662390277
auc: 0.90377
nmi: 0.6153664401847931

epoch 17

0.10012445625750233
recall@1: 0.5232950708980418
recall@2: 0.6453409858203917
recall@4: 0.761647535449021
recall@8: 0.8460499662390277
auc: 0.90254
nmi: 0.6127964836472385

epoch 18

0.09883353459403907
recall@1: 0.5163740715732613
recall@2: 0.6439905469277515
recall@4: 0.7601282916948008
recall@8: 0.8507765023632681
auc: 0.90253
nmi: 0.6128763661848183

epoch 19

0.09817209485722302
recall@1: 0.5300472653612424
recall@2: 0.6541188386225524
recall@4: 0.7648548278190412
recall@8: 0.849594868332208
auc: 0.90627
nmi: 0.610912800294117

epoch 20

0.09693863292228785
recall@1: 0.5286968264686023
recall@2: 0.6482106684672518
recall@4: 0.761647535449021
recall@8: 0.8516205266711682
auc: 0.90406
nmi: 0.6146328977189386

No importance sampling

epoch 0

0.9226418601666874
recall@1: 0.33541525995948684
recall@2: 0.45830519918973667
recall@4: 0.5864280891289669
recall@8: 0.724848075624578
auc: 0.69342
nmi: 0.45570140798912906

epoch 1

0.24120667891572378
recall@1: 0.2635043889264011
recall@2: 0.37660364618501013
recall@4: 0.5042201215395004
recall@8: 0.6453409858203917
auc: 0.72375
nmi: 0.4005879113200249

epoch 2

0.15280496538192012
recall@1: 0.36276164753544904
recall@2: 0.4959486833220797
recall@4: 0.62542201215395
recall@8: 0.7557393652937204
auc: 0.82824
nmi: 0.5016861771823273

epoch 3

0.15254011994552122
recall@1: 0.4186360567184335
recall@2: 0.5430452397029034
recall@4: 0.6764010803511141
recall@8: 0.7895003376097232
auc: 0.8699
nmi: 0.5402947986113584

epoch 4

0.1514671556140021
recall@1: 0.43602295746117486
recall@2: 0.5722484807562458
recall@4: 0.6958136394328157
recall@8: 0.8109385550303849
auc: 0.88091
nmi: 0.565350081686678

epoch 5

0.15368581958665864
recall@1: 0.4523970290344362
recall@2: 0.5810263335584065
recall@4: 0.7106684672518568
recall@8: 0.8144834571235652
auc: 0.88734
nmi: 0.5647634555786827

epoch 6

0.1535880021122704
recall@1: 0.4756920999324781
recall@2: 0.6056718433490884
recall@4: 0.7221471978392978
recall@8: 0.8237677245104659
auc: 0.89771
nmi: 0.5796770339060138

epoch 7

0.14774518743090456
recall@1: 0.47214719783929776
recall@2: 0.6028021607022283
recall@4: 0.7189399054692776
recall@8: 0.8202228224172856
auc: 0.90106
nmi: 0.5788426786170653

epoch 8

0.150999097848328
recall@1: 0.4800810263335584
recall@2: 0.6034773801485482
recall@4: 0.7199527346387576
recall@8: 0.8229237002025659
auc: 0.90103
nmi: 0.5808817682712213

epoch 9

0.15097545799081885
recall@1: 0.4910533423362593
recall@2: 0.6183322079675895
recall@4: 0.7288993923024983
recall@8: 0.826130992572586
auc: 0.90097
nmi: 0.5905139072645302

epoch 10

0.1507523218535146
recall@1: 0.4888588791357191
recall@2: 0.6071910871033086
recall@4: 0.7268737339635382
recall@8: 0.824611748818366
auc: 0.89799
nmi: 0.5830510967034845

epoch 11

0.14282204871961537
recall@1: 0.48700202565833894
recall@2: 0.6147873058744092
recall@4: 0.7314314652261985
recall@8: 0.8273126266036462
auc: 0.89806
nmi: 0.5881259206584548

epoch 12

0.15367429199437674
recall@1: 0.49645509790681974
recall@2: 0.6093855503038488
recall@4: 0.7300810263335584
recall@8: 0.8207292370020257
auc: 0.90591
nmi: 0.5897374001425126

epoch 13

0.14651862120245926
recall@1: 0.49476704929101956
recall@2: 0.6156313301823092
recall@4: 0.7302498311951384
recall@8: 0.8234301147873059
auc: 0.90295
nmi: 0.5872992357178762

epoch 14

0.13935245118714726
recall@1: 0.48970290344361916
recall@2: 0.6154625253207292
recall@4: 0.7258609047940581
recall@8: 0.8192099932478055
auc: 0.90494
nmi: 0.5889746688913337

epoch 15

0.14510860965406883
recall@1: 0.4924037812288994
recall@2: 0.6190074274139096
recall@4: 0.7295746117488183
recall@8: 0.8215732613099257
auc: 0.90142
nmi: 0.5897387978601745

epoch 16

0.1435853014998756
recall@1: 0.48683322079675895
recall@2: 0.6147873058744092
recall@4: 0.7263673193787981
recall@8: 0.8205604321404456
auc: 0.90242
nmi: 0.5872276770777529

epoch 17

0.1343749952543101
recall@1: 0.49358541525995947
recall@2: 0.6147873058744092
recall@4: 0.7312626603646185
recall@8: 0.8232613099257259
auc: 0.906
nmi: 0.5825051329906403

epoch 18

0.13871229041442412
recall@1: 0.48058744091829847
recall@2: 0.6068534773801485
recall@4: 0.7260297096556381
recall@8: 0.8170155300472653
auc: 0.90149
nmi: 0.5851612906457581

epoch 19

0.13462295042911984
recall@1: 0.48548278190411887
recall@2: 0.6063470627954085
recall@4: 0.724341661039838
recall@8: 0.8183659689399054
auc: 0.89962
nmi: 0.5831645930933992

epoch 20

0.13631506395971085
recall@1: 0.4902093180283592
recall@2: 0.6048278190411884
recall@4: 0.7191087103308575
recall@8: 0.8171843349088453
auc: 0.89785
nmi: 0.582648498212415
